src/rdkafka.c

```diff
@@ -270,7 +279,463 @@ void rd_kafka_log0 (const rd_kafka_conf_t *conf,
+                const char value_end_char, char **value, char *errstr, size_t errstr_size) {
+        if (*value) {
+                rd_snprintf(errstr, errstr_size,
+                        "Invalid sasl.oauthbearer.config: multiple '%s' entries", prefix);
+                return;
+        }
+        *loc += strlen(prefix);
+        *value = *loc;
+        while (**loc != '\0' && **loc != value_end_char) {
+                ++*loc;
+        }
+        if (**loc == value_end_char) {
+                // end the string and skip the character
+                **loc = '\0';
+                ++*loc;
+        }
+        // return new allocated memory
+        *value = strdup(*value);
+}
+
+/*
+ * Parse Unsecured JWS config, allocates strings that must be freed
+ */
+static int parse_unsecured_jws_config(const char *cfg,
+                char **principal_claim_name,
+                char **principal,
+                char **scope_claim_name,
+                char **scope_csv_text,
+                int *life_seconds,
+                rd_list_t *extensions, /* rd_strtup_t list */
+                char *errstr, size_t errstr_size) {
+        /*
+         * Extensions:
+         * 
+         * https://tools.ietf.org/html/rfc7628#section-3.1
+         * key            = 1*(ALPHA)
+         * value          = *(VCHAR / SP / HTAB / CR / LF )
+         * 
+         * https://tools.ietf.org/html/rfc5234#appendix-B.1
+         * ALPHA          =  %x41-5A / %x61-7A   ; A-Z / a-z
+         * VCHAR          =  %x21-7E  ; visible (printing) characters
+         * SP             =  %x20     ; space
+         * HTAB           =  %x09     ; horizontal tab
+         * CR             =  %x0D     ; carriage return
+         * LF             =  %x0A     ; linefeed
+         */
+
+        const char *prefix_principal_claim_name = "principalClaimName=";
+        const char *prefix_principal = "principal=";
+        const char *prefix_scope_claim_name = "scopeClaimName=";
+        const char *prefix_scope = "scope=";
+        const char *prefix_life_seconds = "lifeSeconds=";
+        const char *prefix_extension = "extension_";
+
+        char *life_seconds_text = NULL;
+
+        char *cfg_copy = rd_strdup(cfg);
+        char *loc = cfg_copy;
+
+        *principal_claim_name = NULL;
+        *principal = NULL;
+        *scope_claim_name = NULL;
+        *scope_csv_text = NULL;
+        *life_seconds = 0;
+
+        while (*loc != '\0' && *errstr == '\0') {
+                if (*loc == ' ') {
+                        ++loc;
+                } else if (!strncmp(prefix_principal_claim_name, loc, strlen(prefix_principal_claim_name))) {
+                        parse_unsecured_jws_config_value_for_prefix(&loc, prefix_principal_claim_name, ' ',
+                                principal_claim_name, errstr, errstr_size);
+                        if (*errstr == '\0' && **principal_claim_name == '\0') {
```

Comment:
This assumes the errstr is initialized properly, which is not needed if an error return is used instead to indicate errors.

Reply:
👍 Done